pkg/runtime/runtime.go

```diff
@@ -2591,78 +1344,41 @@ func (a *DaprRuntime) stopActor() {
-	}
-	for name, component := range a.compStore.ListStateStores() {
-		a.compStore.DeleteStateStore(name)
-		merr = errors.Join(merr, closeComponent(component, "state store "+name))
-	}
-	for name, component := range a.compStore.ListLocks() {
-		a.compStore.DeleteLock(name)
-		merr = errors.Join(merr, closeComponent(component, "clock store "+name))
-	}
-	for name, component := range a.compStore.ListConfigurations() {
-		a.compStore.DeleteConfiguration(name)
-		merr = errors.Join(merr, closeComponent(component, "configuration store "+name))
-	}
-	for name, component := range a.compStore.ListCryptoProviders() {
-		a.compStore.DeleteCryptoProvider(name)
-		merr = errors.Join(merr, closeComponent(component, "crypto provider "+name))
-	}
-	for name, component := range a.compStore.ListWorkflows() {
-		a.compStore.DeleteWorkflow(name)
-		merr = errors.Join(merr, closeComponent(component, "workflow component "+name))
-	}
-	// Close output bindings
-	// Input bindings are closed when a.ctx is canceled
-	for name, component := range a.compStore.ListOutputBindings() {
-		a.compStore.DeleteOutputBinding(name)
-		merr = errors.Join(merr, closeComponent(component, "output binding "+name))
-	}
-	// Close pubsub publisher
-	// The subscriber part is closed when a.ctx is canceled
-	for name, pubSub := range a.compStore.ListPubSubs() {
-		a.compStore.DeletePubSub(name)
-		if pubSub.Component == nil {
-			continue
+func (a *DaprRuntime) shutdownOutputComponents(ctx context.Context) error {
+	closers := []concurrency.Runner{func(ctx context.Context) error {
+		log.Info("Shutting down name resolver")
+		if closer, ok := a.nameResolver.(io.Closer); ok && closer != nil {
+			if err := closer.Close(); err != nil {
+				err = fmt.Errorf("error closing name resolver: %w", err)
+				log.Warn(err)
+				return err
+			}
 		}
-		merr = errors.Join(merr, closeComponent(pubSub.Component, "pubsub "+name))
-	}
-	merr = errors.Join(merr, closeComponent(a.nameResolver, "name resolver"))
-
-	for _, component := range a.compStore.ListComponents() {
-		a.compStore.DeleteComponent(component.Spec.Type, component.Name)
-	}
-
-	return merr
-}
+		return nil
+	}}
 
-func closeComponent(component any, logmsg string) error {
-	if closer, ok := component.(io.Closer); ok && closer != nil {
-		if err := closer.Close(); err != nil {
-			err = fmt.Errorf("error closing %s: %w", logmsg, err)
-			log.Warn(err)
-			return err
+	closeComponent := func(comp componentsV1alpha1.Component) func(context.Context) error {
+		return func(context.Context) error {
+			log.Infof("Shutting down component %s", comp.LogName())
+			if err := a.processor.Close(comp); err != nil {
+				return err
+			}
+			return nil
 		}
 	}
+	for _, comp := range a.compStore.ListComponents() {
+		closers = append(closers, closeComponent(comp))
+	}
 
-	return nil
+	return concurrency.NewRunnerManager(closers...).Run(ctx)
 }
 
 // ShutdownWithWait will gracefully stop runtime and wait outstanding operations.
```

Comment:
Can you add tests specifically for `ShutdownWithWait` to ensure the order of operations and graceful shutdown period are executed correctly?

Reply:
Sure thing, I'll add those tests! I would like to point out though that the [follow-up daprd PR](https://github.com/dapr/dapr/pull/6720) addresses graceful shutdown specifically. Since this PR is already pretty out of hand as is, I didn't want to also include the wait group and channel changes which are added in that branch which will significantly improve the reliability of everything being fully closed before exit (as well as how that is tracked in runtime.go). Although we obviously strive to make the dapr master branch as correct/stable as possible- please keep in mind that the daprd shutdown handling here is going significantly change in the next PR.

Reply:
Ok, I'll accept the tests in the follow up PR.